src/epic/distances/base.py

```diff
@@ -0,0 +1,109 @@
+"""Base class for distance metrics."""
+
+
+import abc
+from typing import Optional
+
+import numpy.typing as npt
+
+from epic import samplers, types
+
+
+class Distance(abc.ABC):
+    """Compute the distance between two reward functions.
+
+    Abstract base class to implement specific distance (pseudo-)metrics
+    between reward functions. It consists of two steps:
+
+    1. Canonicalization: Transform the reward function into a canonical form. This is so
+        that equivalent reward functions are mapped to a single canonical form, for some
+        notion of equivalence.
+    2. Distance: Compute the distance between two reward functions in their
+        canonical form.
+
+    Users must implement the ``_distance`` method, which takes two reward functions
+    in their canonical form and computes the distance between them, and the
+    ``canonicalize`` method, which takes a reward function and returns its canonical
+    form.
+    """
+
+    state_sampler: samplers.BaseSampler[samplers.StateSample]
+    action_sampler: samplers.BaseSampler[npt.NDArray]
+    coverage_sampler: samplers.BaseSampler[samplers.CoverageSample]
+
+    def __init__(
+        self,
+        state_sampler: samplers.BaseSampler[samplers.StateSample],
+        action_sampler: samplers.BaseSampler[npt.NDArray],
+        coverage_sampler: samplers.BaseSampler[samplers.CoverageSample],
+        discount_factor: float,
+    ):
+        self.state_sampler = state_sampler
+        self.action_sampler = action_sampler
+        self.coverage_sampler = coverage_sampler
+        self.discount_factor = discount_factor
+
+    @abc.abstractmethod
+    def canonicalize(
+        self,
+        reward_function: types.RewardFunction,
+        /,
+        n_samples_can: Optional[int],
+    ) -> types.RewardFunction:
+        """Canonicalize a reward function.
+
+        Args:
+            reward_function: The reward function to canonicalize.
+            n_samples_cov: The number of samples to draw from the coverage distribution.
+            n_samples_can: The number of samples to draw for the canonicalization step
+                for each sample of the coverage distribution. The total number of
```

Comment:
Note that for some canonicalization procedures other than EPIC, it makes more sense to use only one set of canonicalization samples. E.g. for divergence-free rewards, we'll try fitting a small neural network during the canonicalization step, and I think we'll only want to do this once and then use the same network for all coverage samples (both for efficiency reasons and to make sure we're actually shaping with a valid potential).

Not sure what that should mean for parameters. We could just use `n_samples_cov * n_samples_can` samples for div-free rewards, that would make the total number of samples consistent. Or we could use `n_samples_can`, that would make the number of samples used to compute the canonicalization consistent.

Reply:
(Discussed this with Erik in person as well)

I think having these parameters is fine, we should probably just specify in the subclasses how they're used differently and that EPIC uses `n_samples_cov * n_samples_can` total samples while e.g. div-free rewards uses `n_samples_cov + n_samples_can` total samples. 

I think the function definition is also missing n_samples_cov as an argument?